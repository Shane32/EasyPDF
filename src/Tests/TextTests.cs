using System.Drawing;
using System.Text;
using Font = Shane32.EasyPDF.Font;

namespace Tests;

public class TextTests
{
    private readonly PDFWriter _writer = new PDFWriter();

    public TextTests()
    {
        _writer.ScaleMode = ScaleModes.Inches;
        _writer.NewPage(System.Drawing.Printing.PaperKind.Letter, false, 1f, 1f);
        _writer.PrepForTests();
    }

    [Fact]
    public void PositioningTests()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 10f);
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.25f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment);
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void PositioningTests_StretchedY()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 5f) { StretchY = 2f };
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.25f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment);
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void PositioningTests_Spacing()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 10f) { LineSpacing = 1.2f };
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.25f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment);
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void PositioningTests_StretchedY_Spacing()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 5f) { StretchY = 2f, LineSpacing = 1.2f };
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.25f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment);
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void PositioningTests_StretchedX()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 10f) { StretchX = 2f };
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.25f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment);
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

    [Fact]
    public void PositioningTests_MultiLine()
    {
        _writer.Font = new Font(StandardFonts.Helvetica, 10f);
        for (int i = 0; i < 12; i++) {
            _writer.TextAlignment = (TextAlignment)i;
            var pos = new PointF(3f, i * 0.5f);
            var r = 0.0625f;
            _writer.MoveTo(pos).Circle(r);
            _writer.MoveTo(pos).OffsetTo(-r, 0f).LineTo(r + r, 0f);
            _writer.MoveTo(pos).OffsetTo(0f, -r).LineTo(0f, r + r);
            _writer.MoveTo(pos).WriteLine("Alignment: " + _writer.TextAlignment + "\nLine 2");
        }

        _writer.ToArray().SaveAsPdf().ToASCIIString().RemoveID().ShouldMatchApproved(o => o.NoDiff());
    }

}
