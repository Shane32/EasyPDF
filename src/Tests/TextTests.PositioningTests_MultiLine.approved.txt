%PDF-1.4
%\xE2\xE3\xCF\xD3
2 0 obj
<</Length 1003/Filter/FlateDecode>>stream
x\x9C\x95\x97Mo\x141\x0C\x86\xEF\xF3+\xE6\x08\x87\x0Eq\xBE&\xE1F\x11 !.\xB4{C\x1CP\xBB\x94\xA2nK\xA1\x12\x7F\x9FLl\xC7\x1E\xD4v\xB2\xA2h\xDD8v<Ow\xFC\xC6\xF7\xC3\xE9npq\x9CC\x1Cw\x97\xC3\xBB\xDD\xF0y\x80\xD1\x94\x7F'0\xCE\xB6\xFC\x98\xF1\xE20\x98\xF1c\xF9\xFFs0\x13\x8C\x7F\x87/_\x8B\xFFr0u\xDF\xD9\x072~_=\xBE\xCDZ3\x85b\x1D\xC8\xB2\x93\xCF\xA3\x85\xB4|\xF8\xE5w\x88\xF5\xF3b\xB0\xE0\xA6\x00\xB4\x08\xB2\x15j\xFC\xE2_\xAC\x13Z\xAD{O8\xC3\x09\xA7\xA8yq\xB9\x1EG\xDB\xA9\x88\x8B\xE1|\xE0\x84\x87V\xDAM]\xA5$\x87\x81+Z\x96\xEF\x17@\xAF\xDE\xC3\x08f\xDC}\x176\xCB\x9Ey\x824\xEE\x0E\xFA\xF9\xCDB\xF1\xC5\x9B\x9B\xEB\xAB\xDB\xC3\xFE\xF6\xE1\xF5\xF8i\xFF\xFDaw\xF7\xEB\xE5\xEE'\xC2}.\x1F\xC4\xC9\xBBG\x13~\xBA\xBE\xDD\x8F\xB6%\xC1\xBA\xCB\x9F\x8D\x1F\xC1%\xD50U\xB3\xC2j\x1AW\xDE\x0E\x98\x85\xC9\xBA\xBA\x8F\xB6;\xA0,\xD5hdq\x19\x0F\xC5\xED\\\x8B\xA0\xD5\x85\xC5\x06\xB7F\x12\\\xD3A\xD7\xE5)\xE4\x1E\xBCo\x8B\xB5\xFF\xDDE\xD8\xA7)\xF9~\xC2\xE5\x05\xE0\x07\x99\xBD"<GbS\x8DF\x98\x96Am\x07\xCC\xC2\x84cV\x84\xE3LY\xAA\xD1\x08\xE3r=\x94\xB6s-BX\x17f\x1B\xE1\x1A\x89\x84k)\x9B\xDF_\xDB\x83\xF7\xF4\xDB\x9F\xFDM\xA1\xD3\x058\xC1dC?`0\xA9=\x08\x80Q\x88\x01,\xD1A\xABAf\x07\xE8\x10\xA0\\\xCC\x19LP\xA0\xC18\xCEU\xAD\x86\x9A\x1CTJ\x10\xD8\x98\xEB|\x90\xCC\x07U1\xF3\xC6p\x04\x8Eem\x11_\x0E\xC9\x8F\xBF\xE2\xFFS\xBF{x\xB8;\xF4\xB5\x08O5\xA2\xC7\xA1{/\x8F\xE3\xA3\x86^\xDE\x0F\x02U-\x81N\x0E\xD0!@\xB9\x1At\x0F\x1A\xFA\xF2\xFEb\xAEj\x09tt\xD0\xF1\xA0\xA0\xD7\\\xFA\xAAJ/\xD0k8A\xAFe=\x07\x1D\xCA;\xE1\xE6\x11\x02t\xF4jl$]\xDD\xDA\xD8\xC9\xA5\x92\xDD\x1C\xD5\xB1!\x89\xEA@\xD2B\x08\x89\x85\x0C-!\x9FD\x0E[\x08P\xAEF~\x9E5\xF99p\xAEj\x09yt\xE0\xF1\x14\xD2\xEAR\xE4WU\x8A:b8\x91O[\x02\xB9\xB0\xC9\x05Qr\x1Dm\x1C\xC9\xF76r\x82\x9F\xEDQ\xCD|)\xBB\xDDA@\x0B\xA6\xF8\xD1j\xF0\xD9\x01:\x84u\x91\xE1\xF3^\xFE`?Z>9(\xE3*\xBD\xD6\xCEu\x95\xA2\x9E\xE8E\xF8\xE8}\x16~\x88S\xC2\xEE\xD0E\xBE\xBF\xC7#{k\xC3Q}\xDE\x06\x11,\x1B\xB4\x94\xDA\xC0Z\x8A\x96\xB0\x0F\xA2\xA6-\x04(Wc\xEF\xB5\xA0Z\xCF\x8A\x8A\x96\xB0\xF7\xA2\xA9\x1C\xD2\xEAR\xECWU\x8A\xAEb8\xB1\x0F[\xCA\xBA|\xF1\xC1\xD7s\xB6[=\xE1\xEFl\xF6\x04?\xE4\xA3\xFA\xBDM"_6k\x91\xB5\x99E\x16-\x81\x9FEd[\x08P\xAE\x06?i\x91\xB5\x89E\x16-\x81\x9FDd9\xA4\xD5\xA5\xE0\xEB*\x93\x88,\x86\x13\xFC\xBC%\xB2P\xBE\x98~\xA9\xFFIIT\xF0\xCF\xAE\xAF~t]\xCE!\xA5i.7R\xE3\x8F\xBB\xA0[\x910g\xB5\xD0:\xCBB\x8BV\x03\xCF\x0E\xD0!@\xB9\xDA=\xDD\xC2\xEA\xA2\xCEB\x8B\x96\xBA\xAA\x8B\xD0rH\xABK\xDD\xD6WUzu_\x17\xA1\xC5\xB2\x9E\x05_\xEE?\xD1\x95\x8DsG\xBB\xAF\xE0;\xBB=\xB3w\xF1\xA8n\xEF\xA2QC\x88\x96Z\x17Yj\xD1\x12\xF6Q\xA4\xB6\x85\x00\xE5j\xEC\x83\x96Z\x17Xj\xD1\x12\xF6A\xA4\x96CZ]zR\xD2U\x8A\xD4b8\xB1\x8F\x9BR[\xBE\x976\xD7\x04=\xE0\xBB\x9B=\xA3/c\xC81\xCD\xDEe5\xFF\xE5\xF5d*\xA3\xE9\x7F\xB3\xA9\x1AN\xB3\x9EN\xB3\x1AO\xF3j>\xCDm@\xCD\xEB\x095\xAB\x115\xEB\x195\xAF\x86\xD4U\x95jL\xCDzN\xDD\x14\xDA2\x81\xE4X\xCF\xD9n\xF6H\xBF\xAF\xD7\x13{_F\x8C\xBE^\xFF\x0F\x94r\xA3\x04
endstream
endobj
4 0 obj
<</Type/Page/MediaBox[0 0 612 792]/Resources<</ProcSet [/PDF /Text /ImageB /ImageC /ImageI]/Font<</F1 1 0 R>>>>/Contents 2 0 R/Parent 3 0 R>>
endobj
1 0 obj
<</Type/Font/Subtype/Type1/BaseFont/Helvetica/Encoding/WinAnsiEncoding>>
endobj
3 0 obj
<</Type/Pages/Count 1/Kids[4 0 R]/ITXT(3.3.2.0)>>
endobj
5 0 obj
<</Type/Catalog/Pages 3 0 R>>
endobj
6 0 obj
<<>>
endobj
xref
0 7
0000000000 65535 f 
0000001243 00000 n 
0000000015 00000 n 
0000001331 00000 n 
0000001086 00000 n 
0000001396 00000 n 
0000001441 00000 n 
trailer
<</Size 7/Root 5 0 R/Info 6 0 